File: cotsSource/org.apache.qpid/java/common/src/main/java/org/apache/qpid/framing/AMQMethodBody.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.framing;

import org.apache.mina.common.ByteBuffer;
import org.apache.qpid.AMQChannelException;
import org.apache.qpid.AMQConnectionException;
import org.apache.qpid.AMQException;
import org.apache.qpid.protocol.AMQConstant;

public interface AMQMethodBody extends AMQBody
{
    public static final byte TYPE = 1;

    /** AMQP version */
    public byte getMajor();

    public byte getMinor();



    /** @return unsigned short */
    public int getClazz();

    /** @return unsigned short */
    public int getMethod();

    public void writeMethodPayload(ByteBuffer buffer);


    public int getSize();

    public void writePayload(ByteBuffer buffer);

    //public abstract void populateMethodBodyFromBuffer(ByteBuffer buffer) throws AMQFrameDecodingException;

    //public void populateFromBuffer(ByteBuffer buffer, long size) throws AMQFrameDecodingException;

    public AMQFrame generateFrame(int channelId);

    public String toString();



    /**
     * Convenience Method to create a channel not found exception
     *
     * @param channelId The channel id that is not found
     *
     * @return new AMQChannelException
     */
    public AMQChannelException getChannelNotFoundException(int channelId);

    public AMQChannelException getChannelException(AMQConstant code, String message);

    public AMQChannelException getChannelException(AMQConstant code, String message, Throwable cause);

    public AMQConnectionException getConnectionException(AMQConstant code, String message);


    public AMQConnectionException getConnectionException(AMQConstant code, String message, Throwable cause);


    public boolean execute(MethodDispatcher methodDispatcher, int channelId) throws AMQException;
}

File: cotsSource/org.apache.qpid/dotnet/Qpid.Common/Framing/AMQMethodBody.cs
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
using System;
using Apache.Qpid.Buffer;

namespace Apache.Qpid.Framing
{
    public abstract class AMQMethodBody : IBody
    {
        public const byte TYPE = 1;
        
        protected abstract uint BodySize
        {
            get;
        }

        protected abstract ushort Clazz
        {
            get;
        }

        protected abstract ushort Method
        {
            get;
        }

        protected abstract void WriteMethodPayload(ByteBuffer buffer);

        public byte BodyType
        {
            get
            {
                return TYPE;
            }
        }

        public uint Size
        {
            get
            {
                return (uint) (2 + 2 + BodySize);
            }
        }

        public void WritePayload(ByteBuffer buffer)
        {
            buffer.Put(Clazz);
            buffer.Put(Method);
            WriteMethodPayload(buffer);
        }

        /// <summary>
        /// Populates the method body by decoding the specified buffer
        /// </summary>
        /// <param name="buffer">The buffer to decode.</param>
        /// <exception cref="AMQFrameDecodingException">If the buffer cannot be decoded</exception>
        protected abstract void PopulateMethodBodyFromBuffer(ByteBuffer buffer);

        /// <summary>
        /// Populates this instance from a buffer of data.
        /// </summary>
        /// <param name="buffer">The buffer.</param>
        /// <param name="size">The size.</param>
        /// <exception cref="AMQFrameDecodingException">If the buffer contains data that cannot be decoded</exception>
        public void PopulateFromBuffer(ByteBuffer buffer, uint size)
        {
            PopulateMethodBodyFromBuffer(buffer);
        }

        public override string ToString()
        {
            return String.Format("{0}{{ Class: {1} Method: {2} }}", GetType().Name, Clazz, Method);
        }
    }
}
